README.rst
Taxonomy
========

.. image:: https://img.shields.io/pypi/v/taxonomy-connector.svg
    :target: https://pypi.org/project/taxonomy-connector/
    :alt: PyPI

.. image:: http://codecov.io/github/edx/taxonomy-connector/coverage.svg?branch=master
    :target: http://codecov.io/github/edx/taxonomy-connector?branch=master
    :alt: Codecov
    
.. image:: https://github.com/edx/taxonomy-connector/workflows/Python%20CI/badge.svg?branch=master
    :target: https://github.com/edx/taxonomy-connector/actions?query=workflow%3A%22Python+CI%22
    :alt: CI

The taxonomy service is a library that can be installed in other edX components
that provides access to third party taxonomy vendors. EMSI is currently the
only vendor available, but others may be integrated as necessary. This service
can communicate with the vendor to get job, skill, and salary data. This service
can also be used to submit data (course descriptions, etc.) to the vendor to
produce potential matches for a skill or job.

After submitting a pull request, please use the Github "Reviewers" widget to add
relevant reviewers and track review process.


Getting Started
---------------

To install ``taxonomy-connector``, for example, in Course Discovery, follow these steps:

#. It is recommended that you clone this repo into a sub-folder of your working directory. Create a sub-folder with name ``src`` if it doesn't already exist.
#. Clone this repository into the ``src`` folder.
#. Go to the shell of the host environment where you want to install this package and run this ``pip install -e /edx/src/taxonomy-connector``
#. Changes made into the taxonomy repository will now be picked up by your host environment.


Notes:

- In order to communicate with EMSI service, you need to set the values of ``client_id`` and ``client_secret``. These values are picked up from the host environment so you need to pass them in ``.yaml`` file of the host environment.
- Also, to make taxonomy work, the host platform must add an implementation of data providers written in ``./taxonomy/providers``
- Taxonomy APIs use throttle rate set in ``DEFAULT_THROTTLE_RATES`` settings by default. Custom Throttle rate can by set by adding ``ScopedRateThrottle`` class in ``DEFAULT_THROTTLE_CLASSES`` settings and ``taxonomy-api-throttle-scope`` key in ``DEFAULT_THROTTLE_RATES``
- For the skill tags to be verified, the management command ``finalize_skill_tags`` needs to be run periodically.
- Also, You can configure the skill tags verification by setting the values of ``MIN_VOTES_FOR_SKILLS`` and ``RATIO_THRESHOLD_FOR_SKILLS`` in the host platform.


.. code-block:: python

    REST_FRAMEWORK = {
        'DEFAULT_THROTTLE_CLASSES': (
            'rest_framework.throttling.UserRateThrottle',
            'rest_framework.throttling.ScopedRateThrottle'
        ),
        'DEFAULT_THROTTLE_RATES': {
            'user': '100/hour',
            'taxonomy-api-throttle-scope': '60/min',  # custom throttle rate for taxonomy api
        },
    }


Developer Notes
~~~~~~~~~~~~~~~

- To run unit tests, create a virtualenv, install the requirements with ``make requirements`` and then run ``make test``
- To update the requirements, run ``make upgrade``
- To run quality checks, run ``make quality``
- Please do not import models directly in course discovery. e:g if you want to import CourseSkills in Discovery, use the utility get_whitelisted_course_skills instead of directly importing it.


Reporting Security Issues
-------------------------

Please do not report security issues in public. Please email security@edx.org.

Getting Help
------------

Have a question about this repository, or about Open edX in general?  Please
refer to this `list of resources`_ if you need any assistance.

.. _list of resources: https://open.edx.org/getting-help


CHANGELOG.rst
Change Log
==========

..
   All enhancements and patches to edx-enterprise will be documented
   in this file.  It adheres to the structure of http://keepachangelog.com/ ,
   but in reStructuredText instead of Markdown (for ease of incorporation into
   Sphinx documentation and the PyPI description).

   This project adheres to Semantic Versioning (http://semver.org/).

.. There should always be an "Unreleased" section for changes pending release.

Unreleased

[1.29.0] - 2022-11-23
---------------------
* Added XBlockSkills and XBlockSkillData models.
* Added related celery tasks, abstract provider, signals and commands.

[1.28.2] - 2022-11-23
---------------------
* Added industry_names field in Algolia serializer.

[1.28.1] - 2022-11-22
---------------------
* Added JobHolderUsernamesAPIView which returns a list of 100 usernames from SkillsQuiz.


[1.28.0] - 2022-11-21
---------------------
* Updated refresh_job_skills command to save industry relation with job and skills.


[1.27.0] - 2022-10-31
---------------------
* Removed industry foreign key from JobSkills table and create a new table IndustryJobSkill.


[1.26.0] - 2022-10-31
---------------------
* Added relation between JobSkill and Industry Table.

[1.25.0] - 2022-10-24
---------------------
* Added JobTopSkillCategoriesAPIView.

[1.24.0] - 2022-10-21
---------------------
* Added a new model to store industry data using NAICS2 codes.

[1.23.1] - 2022-10-13
---------------------
* Do no concatenate if `short_description is `None`.
* Fix CourseSkills update_or_create call.

[1.23.0] - 2022-10-05
---------------------
* Expand course skills tagging to include `title`, `short_description` and `full_description`.

[1.22.5] - 2022-09-16
---------------------
* Fixes product type issue by using ProductTypes choices.

[1.22.4] - 2022-09-14
---------------------
* Updated utils to support program skills.

[1.22.3] - 2022-09-07
---------------------
* Added support to filter Skills by names.

[1.22.2] - 2022-09-06
---------------------
* Register Program associated models on Admin.
* Change verbose name for RefreshProgramSkillsConfig model

[1.22.1] - 2022-08-26
---------------------
* Added id field in JobSerializer for Algolia.

[1.22.0] - 2022-08-22
---------------------
* Added a new model for storing user response for skills quiz.
* Added new REST endpoints for performing CRUD operations on skills quiz.

[1.21.0] - 2022-08-16
---------------------
* feat: add task to update program skills through EMSI api

[1.20.0] - 2022-08-11
---------------------
* feat: add caching to ``utils.get_whitelisted_serialized_skills()``

[1.19.0] - 2022-08-04
---------------------
* feat: add provider and validator for Programs

[1.18.0] - 2022-08-01
---------------------
* feat: add program skill model
* feat: Update SkillSerializer to include Category and Subcategory details.

[1.17.1] - 2022-07-29
---------------------

* feat: use program update signal to call EMSI API

[1.17.0] - 2022-07-15
---------------------

* refactor: Remove EdxRestApiClient usage in taxonomy-connector

[1.16.3] - 2022-06-23
---------------------

* Added handling for None values for median salary from EMSI.

[1.16.2] - 2022-06-22
---------------------

* Fixed error causes by null values returned by EMSI API.

[1.16.1] - 2022-06-15
---------------------

* Fixed API 429 error and updated admin list display for skills for better usability.

[1.16.0] - 2022-06-08
---------------------

* Added category and subcategory for skill.

[1.15.4] - 2022-04-06
---------------------

* fix: Add limit to EMSI API calls

[1.15.3] - 2022-03-11
---------------------

* fix: Check the course description length after encoding

[1.15.2] - 2022-02-18
---------------------

* feat: Added Support for large size course description translation

[1.15.1] - 2022-02-17
---------------------

* fix: Made the median_posting_duration in JobPosting Nullable to avoid errors on jenkins.

[1.15.0] - 2022-02-11
---------------------

* chore: Removed Django22, 30 and 31 support and added support for Django40

[1.14.5] - 2022-02-08
---------------------

* feat: Added Support for course description translation

[1.14.4] - 2022-01-28
---------------------

* feat: Add Translation model

[1.14.3] - 2021-10-27
---------------------

* fix: Make job names unique and handle exception where ever job is created/updated

[1.14.2] - 2021-09-08
---------------------

* Fixed an issue that was causing an error while index jobs data to algolia.

[1.14.1] - 2021-08-20
---------------------

* Decreased skills query chunk_size from 2000 to 50 to fetch more jobs.

[1.14.0] - 2021-08-16
---------------------

* Added managment command and related code to index jobs data to algolia.

[1.13.0] - 2021-08-9
---------------------

* Added Skill, Job and JobPostings viewsets.

[1.12.2] - 2021-08-5
---------------------

* Add job posting information in utility method `get_course_jobs`.

[1.12.1] - 2021-08-3
---------------------

* Add utility method `get_course_jobs` to return job associated with a course.

[1.12.0] - 2021-07-13
---------------------

* Added support for django 3.1 and 3.2

[1.11.2] - 2021-05-28
---------------------

* Added utility method to return serialized course skills.

[1.11.1] - 2021-04-20
---------------------

* Fixed .rst issues in CHANGELOG.rst

[1.11.0] - 2021-04-16
---------------------

* Mention currency in median salary field and add verbose name for models.

[1.10.0] - 2021-04-12
---------------------

* Remove all the usages of old `course_id` field including the column definition in `CourseSkills` model.

[1.9.0] - 2021-04-12
--------------------

* Replace the usages of old `course_id` in `CourseSkills` with the new `course_key` field.

[1.8.0] - 2021-04-09
--------------------

* Added a new field named `course_key` in `CourseSkills` model to deprecate and replace the old `course_id` field in future.

[1.7.0] - 2021-04-07
--------------------

* Removed RefreshCourseSkill view.


taxonomy/constants.py
# -*- coding: utf-8 -*-
"""
Constants used by the taxonomy connector.
"""

from datetime import date

from dateutil.relativedelta import relativedelta
from django.conf import settings


def get_lookup_query_filter(external_ids):
    """
    Build query filter for the lookup endpoint.
    """
    lookup_query_filter = {
        'ids': external_ids
    }
    return lookup_query_filter


def get_job_query_filter(skills=None, industry=None):
    """
    Build job query filter to be used in fetching jobs data from the EMSI Service.

    Arguments:
        skills (list): jobs external ids to add in filter
        industry (Industry): Industry object to add in filter

    Returns:
        dict: Job postings query filter to used in EMSI API
    """
    jobs_query_filter = {
        'filter': {
            'when': {
                'start': str(date.today() - relativedelta(months=6)),
                'end': str(date.today()),
                'type': 'active'
            },
        },
        # we are using TITLE facet for `rank` and SKILLS facet for the `nested rank`
        # for every skill's batch in our system (let say 50 skills) we are getting top 100 jobs by `unique_postings`
        # and 10 skills per every job ranked by significance.
        'rank': {   # jobs
            'by': 'unique_postings',
            'limit': 100  # rank.limit for nested rankings must be <= 100
        },
        'nested_rank': {  # skills
            'by': 'significance',
            'limit': 10  # nested_rank.limit for nested rankings must be <= 1000
        }
    }
    if skills:
        jobs_query_filter['filter']['skills'] = {
            'include': skills,
            'include_op': 'or'
        }
    if industry:
        jobs_query_filter['filter']['naics2'] = {
            'include': [str(industry.code)]
        }
        jobs_query_filter['rank']['limit'] = 25  # for every batch get top 25 jobs
        jobs_query_filter['nested_rank']['limit'] = 10  # for every job get top 10 skills
    return jobs_query_filter


def get_job_posting_query_filter(jobs=None):
    """
    Build job  query filter to be used in fetching jobs posting data from the EMSI Service.

    Arguments:
        jobs (list): jobs external ids to add in filter

    Returns:
        dict: Job postings query filter to used in EMSI API
    """
    job_posting_query_filter = {
        'filter': {
            'when': {
                'start': str(date.today() - relativedelta(months=6)),
                'end': str(date.today()),
                'type': 'active'
            }
        },
        # We are using TITLE facet for `rank` and for every job batch in our system (let say 100 job)
        # we are getting jobpostings by `unique_postings` with max limit to 1000.
        # There is always one JobPosting against each job we pass to the API.
        'rank': {
            'by': 'unique_postings',
            'limit': 1000,   # rank.limit must be <= 1000
            'extra_metrics': [
                'median_posting_duration',
                'median_salary',
                'unique_companies'
            ]
        }
    }
    if jobs:
        job_posting_query_filter['filter']['title'] = {
            'include': jobs,
            'include_op': 'or'
        }
    return job_posting_query_filter


AMAZON_TRANSLATION_ALLOWED_SIZE = 5000
EMSI_API_RATE_LIMIT_PER_SEC = 5
TRANSLATE_SERVICE = 'translate'
ENGLISH = 'en'
AUTO = 'auto'
REGION = 'us-east-1'

NAICS2_CODES = {
    11: 'Agriculture, Forest, Fishing and Hunting',
    21: 'Mining',
    22: 'Utilities',
    23: 'Construction',
    31: 'Manufacturing',
    32: 'Manufacturing',
    33: 'Manufacturing',
    42: 'Wholesale Trade',
    44: 'Retail Trade',
    45: 'Retail Trade',
    48: 'Transportation and Warehousing',
    49: 'Transportation and Warehousing',
    51: 'Information',
    52: 'Finance and Insurance',
    53: 'Real Estate and Rental and Leasing',
    54: 'Professional, Scientific, and Technical Services',
    55: 'Management of Companies and Enterprises',
    56: 'Administrative and Support and Waste Management and Redmediation Services',
    61: 'Educational Services',
    62: 'Health Care and Social Assistance',
    71: 'Arts, Entertainment, and Recreation',
    72: 'Accommodation and Food Services',
    81: 'Other Services (except Public Administration)',
    92: 'Public Administration',
}

MIN_VOTES_FOR_SKILLS = settings.MIN_VOTES_FOR_SKILLS \
    if hasattr(settings, "MIN_VOTES_FOR_SKILLS") else 0
RATIO_THRESHOLD_FOR_SKILLS = settings.RATIO_THRESHOLD_FOR_SKILLS \
    if hasattr(settings, "RATIO_THRESHOLD_FOR_SKILLS") else 1
